File: others/string-concat/lua/bench_strcat.lua
-- -*- coding: utf-8 -*-

---
--- $Release: $
--- $Copyright: copyright(c) 2011 kuwata-lab.com all rights reserved $
--- $License: Creative Commons Attribution (CC BY) $
---

benchmarker = require('benchmarker')

members = {"Haruhi", "Mikuru", "Yuki", "Itsuki", "Kyon"}


expected = null

fh, msg = io.open("../expected.out", "r")
if fh then
	expected = fh:read("*a")
	fh:close()
else
	print('*** error: ' .. msg)
end


function verify(out)
	if expected ~= out then
		print("*** expected == out: failed")
	end
end


loop = 0 + (os.getenv('N') or 100*1000)
debug = os.getenv('D') or false

bm = benchmarker.Benchmarker.new(loop)


function bench_empty(n)
	local m = members
	local s1, s2, s3, s4, s5 = m[1], m[2], m[3], m[4], m[5]
	local out
	for i = 1, n do
		for j = 1, 20 do
			-- pass
		end
	end
end
bm:empty_task(bench_empty)


title = "str..str..str"
function bench_strcat1(n)
	local m = members
	local s1, s2, s3, s4, s5 = m[1], m[2], m[3], m[4], m[5]
	local out
	for i = 1, n do
		buf = ""
		buf = buf .. "<table>\n"
		for j = 1, 20 do
			buf = buf .. "  <tr>\
    <td>" .. s1 .. "</td>\
    <td>" .. s2 .. "</td>\
    <td>" .. s3 .. "</td>\
    <td>" .. s4 .. "</td>\
    <td>" .. s5 .. "</td>\
  </tr>\n"
		end
		buf = buf .. "</table>\n"
	end
	if debug then verify(buf) end
end
bm:task(title, bench_strcat1)


title = "buf[#buf+1]=str;table.concat()"
function bench_strcat2(n)
	local m = members
	local s1, s2, s3, s4, s5 = m[1], m[2], m[3], m[4], m[5]
	local out
	for i = 1, n do
		buf = {}
		buf[#buf+1] = "<table>\n"
		for j = 1, 20 do
			buf[#buf+1] = "  <tr>\
    <td>"; buf[#buf+1] = s1; buf[#buf+1] = "</td>\
    <td>"; buf[#buf+1] = s2; buf[#buf+1] = "</td>\
    <td>"; buf[#buf+1] = s3; buf[#buf+1] = "</td>\
    <td>"; buf[#buf+1] = s4; buf[#buf+1] = "</td>\
    <td>"; buf[#buf+1] = s5; buf[#buf+1] = "</td>\
  </tr>\n";
		end
		buf[#buf+1] = "</table>\n"
		out = table.concat(buf);
	end
	if debug then verify(buf) end
end
bm:task(title, bench_strcat2)


title = "buf[#buf+1]=str only (no concat)"
function bench_strcat3(n)
	local m = members
	local s1, s2, s3, s4, s5 = m[1], m[2], m[3], m[4], m[5]
	local out
	for i = 1, n do
		buf = {}
		buf[#buf+1] = "<table>\n"
		for j = 1, 20 do
			buf[#buf+1] = "  <tr>\
    <td>"; buf[#buf+1] = s1; buf[#buf+1] = "</td>\
    <td>"; buf[#buf+1] = s2; buf[#buf+1] = "</td>\
    <td>"; buf[#buf+1] = s3; buf[#buf+1] = "</td>\
    <td>"; buf[#buf+1] = s4; buf[#buf+1] = "</td>\
    <td>"; buf[#buf+1] = s5; buf[#buf+1] = "</td>\
  </tr>\n";
		end
		buf[#buf+1] = "</table>\n"
		--out = table.concat(buf);
	end
	--if debug then verify(buf) end
end
bm:task(title, bench_strcat3)


title = "table.insert();table.concat()"
function bench_strcat4(n)
	local m = members
	local s1, s2, s3, s4, s5 = m[1], m[2], m[3], m[4], m[5]
	local out
	for i = 1, n do
		buf = {}
		table.insert(buf, "<table>\n")
		for j = 1, 20 do
			table.insert(buf, "  <tr>\
    <td>"); table.insert(buf, s1); table.insert(buf, "</td>\
    <td>"); table.insert(buf, s2); table.insert(buf, "</td>\
    <td>"); table.insert(buf, s3); table.insert(buf, "</td>\
    <td>"); table.insert(buf, s4); table.insert(buf, "</td>\
    <td>"); table.insert(buf, s5); table.insert(buf, "</td>\
  </tr>\n");
		end
		table.insert(buf, "</table>\n")
		out = table.concat(buf);
	end
	if debug then verify(buf) end
end
bm:task(title, bench_strcat4)


title = "table.insert() only (no concat)"
function bench_strcat5(n)
	local m = members
	local s1, s2, s3, s4, s5 = m[1], m[2], m[3], m[4], m[5]
	local out
	for i = 1, n do
		buf = {}
		table.insert(buf, "<table>\n")
		for j = 1, 20 do
			table.insert(buf, "  <tr>\
    <td>"); table.insert(buf, s1); table.insert(buf, "</td>\
    <td>"); table.insert(buf, s2); table.insert(buf, "</td>\
    <td>"); table.insert(buf, s3); table.insert(buf, "</td>\
    <td>"); table.insert(buf, s4); table.insert(buf, "</td>\
    <td>"); table.insert(buf, s5); table.insert(buf, "</td>\
  </tr>\n");
		end
		table.insert(buf, "</table>\n")
		--out = table.concat(buf);
	end
	--if debug then verify(buf) end
end
bm:task(title, bench_strcat5)



loop = 10*1000


function main()
	bm:run()
end


main()

File: others/string-concat/ruby/bench_strcat.rb
# -*- coding: utf-8 -*-

###
### $Release: $
### $Copyright: copyright(c) 2011 kuwata-lab.com all rights reserved $
### $License: Creative Commons Attribution (CC BY) $
###

require './benchmarker'

members = ["Haruhi", "Mikuru", "Yuki", "Itsuki", "Kyon"]
#s1, s2, s3, s4, s5 = members

loop  = (ENV['N'] || 100*1000).to_i
cycle = (ENV['C'] || 3).to_i

$expected = File.open('../expected.out') {|f| f.read }

def verify(out)
  $expected == out or raise "$expected == out: failed"
end

puts "## loop = #{loop}"
puts

Benchmarker.new(:cycle=>cycle) do |bm|

  bm.empty_task do
    for i in 1..loop
      for j in 1..20
        # pass
      end
    end
  end

  out = nil

  bm.task "String#<<" do
    s1, s2, s3, s4, s5 = members
    for i in 1..loop
      buf = ""
      buf << "<table>\n"
      for j in 1..20
        buf << "  <tr>
    <td>" << s1 << "</td>
    <td>" << s2 << "</td>
    <td>" << s3 << "</td>
    <td>" << s4 << "</td>
    <td>" << s5 << "</td>
  </tr>\n"
      end
      buf << "</table>\n"
      out = buf
    end
  end
  verify(out)

  bm.task "String#<< + \#{}" do
    s1, s2, s3, s4, s5 = members
    for i in 1..loop
      buf = ""
#      buf << %Q`<table>\n`
#      for j in 1..20
#        buf << %Q`  <tr>
#    <td>#{s1}</td>
#    <td>#{s2}</td>
#    <td>#{s3}</td>
#    <td>#{s4}</td>
#    <td>#{s5}</td>
#  </tr>\n`
#      end
#      buf << %Q`</table>\n`
      buf << "<table>\n"
      for j in 1..20
        buf << "  <tr>
    <td>#{s1}</td>
    <td>#{s2}</td>
    <td>#{s3}</td>
    <td>#{s4}</td>
    <td>#{s5}</td>
  </tr>\n"
      end
      buf << "</table>\n"
      out = buf
    end
  end
  verify(out)

  bm.task "Array#push + join" do
    s1, s2, s3, s4, s5 = members
    for i in 1..loop
      buf = []
      buf.push("<table>\n")
      for j in 1..20
        buf.push("  <tr>
    <td>", s1, "</td>
    <td>", s2, "</td>
    <td>", s3, "</td>
    <td>", s4, "</td>
    <td>", s5, "</td>
  </tr>\n")
      end
      buf.push("</table>\n")
      out = buf.join()
    end
  end
  verify(out)

  bm.task "Array#<< + join" do
    s1, s2, s3, s4, s5 = members
    for i in 1..loop
      buf = []
      buf << "<table>\n"
      for j in 1..20
        buf << "  <tr>
    <td>" << s1 << "</td>
    <td>" << s2 << "</td>
    <td>" << s3 << "</td>
    <td>" << s4 << "</td>
    <td>" << s5 << "</td>
  </tr>\n"
      end
      buf << "</table>\n"
      out = buf.join()
    end
  end
  verify(out)

  bm.task "Array#push only (no join)" do
    s1, s2, s3, s4, s5 = members
    for i in 1..loop
      buf = []
      buf.push("<table>\n")
      for j in 1..20
        buf.push("  <tr>
    <td>", s1, "</td>
    <td>", s2, "</td>
    <td>", s3, "</td>
    <td>", s4, "</td>
    <td>", s5, "</td>
  </tr>\n")
      end
      buf.push("</table>\n")
      #out = buf.join()
    end
  end
  #verify(out)

end
